sentry-android-okhttp/src/main/java/io/sentry/android/okhttp/SentryOkHttpInterceptor.kt

```diff
@@ -104,6 +117,123 @@ class SentryOkHttpInterceptor(
         }
     }
 
+    // TODO: ignore exceptions of the type UnknownHostException
+
+    private fun captureEvent(request: Request, response: Response) {
+        // not possible to get a parameterized url, but we remove at least the
+        // query string and the fragment.
+        // url example: https://api.github.com/users/getsentry/repos/#fragment?query=query
+        // url will be: https://api.github.com/users/getsentry/repos/
+        // ideally we'd like a parameterized url: https://api.github.com/users/{user}/repos/
+        // but that's not possible
+        var requestUrl = request.url.toString()
+
+        val query = request.url.query
+        if (!query.isNullOrEmpty()) {
+            requestUrl = requestUrl.replace("?$query", "")
+        }
+
+        val urlFragment = request.url.fragment
+        if (!urlFragment.isNullOrEmpty()) {
+            requestUrl = requestUrl.replace("#$urlFragment", "")
+        }
+
+        if (!captureFailedRequests || !TracePropagationTargets.contain(failedRequestsTargets, requestUrl) || !containsStatusCode(response.code)) {
+            return
+        }
+
+        val mechanism = Mechanism().apply {
+            type = "SentryOkHttpInterceptor"
+
+//            // with description, mechanism in the UI is buggy
+//            description = message
+
+            // TODO: should it be synthetic? likely not, mechanism SentryOkHttpInterceptor should be considered for grouping
+//            synthetic = true
+        }
+        val exception = SentryHttpClientError("Event was captured because the request status code was ${response.code}")
+        // TODO: okhttp thread uses thread pool so its always the same stack trace (okhttp frames only), stacktrace is set as snapshot=true, is that ok?
```

Comment:
@untitaker Do you know if those fields should be marked as `synthetic` or `snapshot`?

For context:
We're raising an event out of an HTTP error.
The thread info isn't so relevant because it comes from the okhttp thread pool, not super useful, but the event has enough info in the `request` and `response` fields.
The stack trace is a snapshot from that thread instead of the stack trace of an exception (since its a synthetic exception).

Reply:
Or, a more general question, Events created out of HTTP errors, what's the best way to group them?

Reply:
@marandaneto I honestly can't answer the general question here definitively, even customers have conflicting opinions about how they should be grouped.

* `synthetic` means: ignore the exception type as it is not stable across platforms. I doubt this applies here.
* `snapshot` would apply here IMO.

`snapshot` would give you the same grouping as ANR, i.e. by stacktrace (ignoring hierarchical grouping for a sec). But that's just what the protocol defines. I can't tell you whether a customer will walk in and say "well actually, i'd like to group them by http status code"

Reply:
`synthetic` is for crashes caught by signal handlers, so I would omit this flag entirely.

With `synthetic=true`, our grouping ignores the `type` of the exception entirely (see reasoning [here](https://github.com/getsentry/sentry/blob/1e5966cff9cbdcab2160654500b6f33f387ccc4e/src/sentry/grouping/strategies/newstyle.py#L597-L606)).

It seems to me though that since a bad response from a server is _not_ an exception, maybe we should not use the `exception` interface at all, and use the [`threads`](https://develop.sentry.dev/sdk/event-payloads/types/#thread) interface instead?

Settings `snapshot` should be OK, though it does [invert the hierarchy](https://github.com/getsentry/sentry/blob/1e5966cff9cbdcab2160654500b6f33f387ccc4e/src/sentry/grouping/strategies/newstyle.py#L524-L525) for hierarchical grouping - unclear what the consequences of that are in this case.

Reply:
Ok not setting `synthetic`.
Setting `snapshot`.
@jjbayer some 3rd party libs do throw an exception when there's a 4xx or 5xx, in this case, OkHttp doesn't, but since we have more integrations, I'd prefer to have them aligned with a sort of fake exception that it has its current stack trace.